packages/compiler/emitter-framework/types.ts

```diff
@@ -0,0 +1,148 @@
+import {
+  Enum,
+  Interface,
+  Model,
+  ModelProperty,
+  Operation,
+  Program,
+  Scalar,
+  Tuple,
+  Type,
+  Union,
+} from "../core/index.js";
+import { Placeholder } from "./placeholder.js";
+
+export interface AssetEmitter<T> {
+  /**
+   * Get the current emitter context as set by the TypeEmitter's various
+   * context methods.
+   *
+   * @returns The current emitter context
+   */
+  getContext(): Context;
+  getOptions(): Record<string, unknown>;
+  getProgram(): Program;
+  emitTypeReference(type: Type): EmitEntity<T>;
+  emitDeclarationName(type: CadlDeclaration): string;
+  emitType(type: Type): EmitEntity<T>;
+  emitProgram(options?: { emitGlobalNamespace?: boolean; emitCadlNamespace?: boolean }): void;
+  emitModelProperties(model: Model): EmitEntity<T>;
+  emitModelProperty(prop: ModelProperty): EmitEntity<T>;
+  emitOperationParameters(operation: Operation): EmitEntity<T>;
+  emitOperationReturnType(operation: Operation): EmitEntity<T>;
+  emitInterfaceOperations(iface: Interface): EmitEntity<T>;
+  emitInterfaceOperation(operation: Operation): EmitEntity<T>;
+  emitEnumMembers(en: Enum): EmitEntity<T>;
+  emitUnionVariants(union: Union): EmitEntity<T>;
+  emitTupleLiteralValues(tuple: Tuple): EmitEntity<T>;
+  createSourceFile(name: string): SourceFile<T>;
+  createScope(sourceFile: SourceFile<T>, name: string): SourceFileScope<T>;
+  createScope(namespace: any, name: string, parentScope: Scope<T>): NamespaceScope<T>;
+  createScope(block: any, name: string, parentScope?: Scope<T> | null): Scope<T>;
+  result: {
+    declaration(name: string, value: T | Placeholder<T>): Declaration<T>;
+    rawCode(value: T | Placeholder<T>): RawCode<T>;
+    none(): NoEmit;
+  };
+  writeOutput(): Promise<void>;
+}
+
+export interface ScopeBase<T> {
+  kind: string;
+  name: string;
+  parentScope: Scope<T> | null;
+  childScopes: Scope<T>[];
+  declarations: Declaration<T>[];
+}
+
+export interface SourceFileScope<T> extends ScopeBase<T> {
+  kind: "sourceFile";
+  sourceFile: SourceFile<T>;
+}
+
+export interface NamespaceScope<T> extends ScopeBase<T> {
+  kind: "namespace";
+  parentScope: Scope<T>;
+  namespace: any;
+}
+
+export type Scope<T> = SourceFileScope<T> | NamespaceScope<T>;
+
+export interface TypeReference {
+  expression: string;
+}
+
+export interface SourceFile<T> {
```

Comment:
Is it going to cause problems that there is also a SourceFile core type? Should we pick another name?

Reply:
Yeah, I'll rename to like EmittedSourceFile or so.

Reply:
(Technically it won't cause a conflict since the emitter framework is exported under a separate entrypoint, but better not to cause confusion).